src/Cardano/Shell/NodeIPC/ServerExample.hs

```diff
@@ -0,0 +1,128 @@
+import           GHC.IO.Handle.FD (fdToHandle)
+import           Prelude (String)
+import           System.Environment (lookupEnv, setEnv, unsetEnv)
+import           System.IO (BufferMode (..), hClose, hSetBuffering)
+import           System.Process (CreateProcess (..), StdStream (..), createPipe,
+                                 createPipeFd, proc, withCreateProcess)
+
+-- | Create a pipe for interprocess communication and return a
+-- ('ReadHandle', 'WriteHandle') Handle pair.
+getReadWriteHandles :: IO (ReadHandle, WriteHandle)
+getReadWriteHandles = do
+    (readHndl, writeHndl) <- createPipe
+    hSetBuffering readHndl LineBuffering
+    hSetBuffering writeHndl LineBuffering
+    let readHandle  = ReadHandle readHndl
+    let writeHandle = WriteHandle writeHndl
+    return (readHandle, writeHandle)
+
+--------------------------------------------------------------------------------
+-- Testing
+--------------------------------------------------------------------------------
+
+nodePort :: Port
+nodePort = Port 8090
+
+-- | Example using file descriptor
+exampleWithFD :: MsgIn -> IO (MsgOut, MsgOut)
+exampleWithFD msgin = do
+
+    (serverReadHandle, serverWriteHandle) <- getReadWriteHandles
+
+    (_, responses) <- do
+        (clientReadHandle, clientWriteHandle) <- getReadWriteHandles
+        -- Send message to client
+        sendMessage clientWriteHandle msgin
+        startIPC SingleMessage clientReadHandle serverWriteHandle nodePort
+        `concurrently`
+        receieveMessages serverReadHandle
+
+    return responses
+
+-- | Example of an IPC server that is using haskell executable as an server.
+--
+-- This will be the server, the one which sends the message (such as @Ping@, @QueryPort@)
+-- to get the response from the client.
+-- The client is executed via @stack exec node-ipc haskell@
+exampleServerWithProcess :: MsgIn -> IO (MsgOut, MsgOut)
+exampleServerWithProcess msg = bracket acquire restore (action msg)
+  where
+    acquire :: IO (ReadHandle, Handle)
+    acquire = do
+        (rFd, wFd) <- createPipeFd
+        -- Set the write file descriptor to the envrionment variable
+        -- the client will look this up, and use it to talk the server
+        setEnv "NODE_CHANNEL_FD" (show wFd)
+        readHandle  <- ReadHandle <$> fdToHandle rFd
+        -- Since closeFd only exists in 'unix' library,
+        -- the only way to close this Fd is to convert it into Handle and apply
+        -- hClose to it
+        writeHandle <- fdToHandle wFd
+        return (readHandle, writeHandle)
+
+    restore :: (ReadHandle, Handle) -> IO ()
+    restore ((ReadHandle rHandle), wHandle) = do
+        hClose rHandle
+        hClose wHandle
+        unsetEnv "NODE_CHANNEL_FD"
+
+    action :: MsgIn -> (ReadHandle, Handle) -> IO (MsgOut, MsgOut)
+    action msgin (readHandle, _) = do
+        withCreateProcess (proc "stack" ["exec", "node-ipc", "haskell"])
+            { std_in = CreatePipe } $
+                \(Just stdIn) _ _ _ -> do
```

Comment:
I don't like this since it's not really in sync with the JS side, but fine for now I guess.